src/Terrain/Maps/MapReader.java

```diff
@@ -0,0 +1,171 @@
+            // System.out.println();
+            String mapName = fileEntry.getName();
+            if( !mapName.contains(".map") )
+              continue;
+            mapName = mapName.replaceAll("_", " ");
+            mapName = mapName.replaceAll(".map", "");
+            System.out.println("Parsing map: " + mapName);
+            Scanner scanner = new Scanner(fileEntry);
+            String line = scanner.nextLine();
+            ArrayList<ArrayList<Environment.Terrains>> terrainData = new ArrayList<ArrayList<Environment.Terrains>>();
+
+            ArrayList<ArrayList<XYCoord>> properties = new ArrayList<ArrayList<XYCoord>>(4);
+            properties.add(new ArrayList<XYCoord>());
+            properties.add(new ArrayList<XYCoord>());
+            properties.add(new ArrayList<XYCoord>());
+            properties.add(new ArrayList<XYCoord>());
+
+            for( int i = 0; i < line.length() / 4; i++ )
+            {
+              terrainData.add(new ArrayList<Environment.Terrains>());
+            }
+            for( int yCoord = 0; scanner.hasNextLine(); line = scanner.nextLine() )
+            {
+              for( int i = 0; i < line.length(); i += 4 )
+              {
+                int sideNumber = 99;
+                try
+                {
+                  sideNumber = Integer.parseInt(("" + line.charAt(i) + line.charAt(i + 1)).trim());
+                }
+                catch (NumberFormatException e)
+                {}
+                if( sideNumber < 4 ) // TODO: add more sides
+                {
+                  properties.get(sideNumber).add(new XYCoord(i / 4, yCoord));
+                }
+                String terrainCode = "" + line.charAt(i + 2) + line.charAt(i + 3);
+                terrainData.get(i / 4).add(stringToCode(terrainCode));
+              }
+              yCoord++;
+            }
+            ArrayList<Environment.Terrains[]> terrainArrayArray = new ArrayList<Environment.Terrains[]>();
+            for( int i = 0; i < terrainData.size(); i++ )
+            {
+              terrainArrayArray.add(terrainData.get(i).toArray(new Environment.Terrains[0]));
+            }
+            ArrayList<XYCoord[]> propertyArrayArray = new ArrayList<XYCoord[]>();
+            for( int i = 0; i < properties.size(); i++ )
+            {
+              if( properties.get(i).size() > 0 )
+                propertyArrayArray.add(properties.get(i).toArray(new XYCoord[0]));
+            }
+            MapInfo info = new MapInfo(mapName, terrainArrayArray.toArray(new Environment.Terrains[0][0]),
+                propertyArrayArray.toArray(new XYCoord[0][0]));
+            importMaps.add(info);
+            scanner.close();
+          }
+          // skip the file if you can't find it
+          catch (FileNotFoundException e)
+          {
+            System.out.println("WARNING: Could not find map file " + fileEntry.getName());
+          }
+        }
+      }
+    }
+    catch (NullPointerException e)
+    {}
```

Comment:
Another empty `catch` block. Gross.

Reply:
I don't even know what *could* trigger that catch block, unless the operating system is generally borked in some way.
But Java would scream and wail and carry on if I didn't have it.